lib/trie/trie.go

```diff
@@ -26,6 +26,11 @@ type Trie struct {
 	// pruner to detect with database keys (trie node Merkle values) can
 	// be deleted.
 	deletedMerkleValues map[string]struct{}
+	// newDeletedMerkleValues contains deleted Merkle values for operations
+	// in progress. Its goal is to prevent having the trie `deletedMerkleValues`
+	// in a bad state if one of the trie operation modified some (deep copied) trie
+	// node and then failed.
+	newDeletedMerkleValues map[string]struct{}
```

Comment:
is there a need for this? Could you not just lock and unlock the `trie` so it will never be in a bad state.

Reply:
This is about tracking deleted merkle values in the trie since the last trie snapshot, not the actual trie data.
The trie data is unaffected if an error occurs (since we don't update the root node of the trie), but those tracked merkle values would be wrongly tracked in case of an error occuring in the middle of an operation.

Reply:
I think every trie method should be transactional; the transaction mechanism in the state (based on the trie snapshot) allows to run multiple method calls and rollback (the trie data for now, not the tracked stuff, although this could be arranged).

However, there is a lot of trie calls without snapshotting the trie, and in that case, if the operation fails in the middle (for now it's only `LoadFromMap`, but soon all operations may fail at any time during their recursion), the trie will be left in a bad state.

Finally - that came up today - I actually need to track the deleted (and inserted) Merkle values during a single operation to track the inserted Merkle values correctly (i.e. a temporarily inserted node which then gets thrown out in the same method call should not be tracked as inserted).